marconi-core-json-rpc/src/Marconi/Core/JsonRpc.hs

```diff
@@ -100,6 +101,24 @@ mkHttpRequestTracer trace =
         , outputFormat = DetailedWithSettings def{mPrelogRequests = True}
         }
 
+{- | Given two mapping functions, one for the query and one for the result,
+convert a query handler into one that takes a modified version of the query
+and returns a modified version of the result.
+
+This is technically a profunctor operation with some extra levels of @Functor@
+on the right: @dimapHandler l ≡ dimap l . fmap . fmap@
+
+However, since we need it only for @(->)@ it's not worth pulling in the `profunctor`
+dependency.
+-}
+dimapHandler
+  :: (Functor m)
+  => (qry' -> qry)
+  -> (res -> res')
```

Comment:
I think there are some case where we may need an effectful result here `res -> m res'` (I'm thinking about Wil's current work to add a stability flag to the result, that requires access to `lastStablePoint`, which is effectful).
We can either provide a monadic and a pure variant or just a effectful version, both sounds okay to me.

This aside, I think the design is great.

Reply:
It seems to me that if we need an effectful modification of either the query or the result, we're no longer looking at the simple adaptation of one handler into another. In that situation, I think it would be clearer to write the code more directly. Would you be OK with adding an effectful mapping function later if it turns out to be worthwhile?